flash-patcher.py

```diff
@@ -24,111 +15,147 @@
-        JPEXS_PATH = path
-        JPEXS_ARGS = args
-        return True
-
-    return False
+import argparse
+import base64
+import os
+import shutil
+import subprocess
+import sys
+from logging import basicConfig, critical, error, exception, info
+from pathlib import Path
 
-"""
-Detect JPEXS install location. Returns True if successful.
-"""
-def detect_jpexs():
-    # apt install location
-    if (set_jpexs_if_exists("/usr/bin/ffdec")):
-        return True
+basicConfig(level=1, format="%(levelname)s: %(message)s")
 
+CURRENT_VERSION = "v4.1.5"
 
-    # flatpak install location
-    if (set_jpexs_if_exists("/usr/bin/flatpak", ["run", "--branch=stable", "--arch=x86_64", "--command=ffdec.sh", "com.jpexs.decompiler.flash"])):
-        # Detect if JPEXS is installed. The function call can only detect if Flatpak is installed
-        testrun = subprocess.run([JPEXS_PATH] + JPEXS_ARGS + ["-help"], stdout=subprocess.DEVNULL)
+DECOMP_LOCATION = Path("./.Patcher-Temp/mod/")
+DECOMP_LOCATION_WITH_SCRIPTS = Path(DECOMP_LOCATION, "scripts/")
 
-        if (testrun.returncode == 0):
-            return True
+LOCATION_APT = Path("/usr/bin/ffdec")
+LOCATION_FLATPAK = Path("/usr/bin/flatpak")
+LOCATION_WINDOWS = Path(f"{os.getenv('PROGRAMFILES')}\\FFDec\\ffdec.exe")
+LOCATION_WOW64 = Path(f"{os.getenv('PROGRAMFILES(X86)')}\\FFDec\\ffdec.exe")
 
-    # windows default install location
-    if (set_jpexs_if_exists(f"{os.getenv('ProgramFiles')}\\FFDec\\ffdec.exe")):
-        return True
+ARGS_FLATPAK = [
+    "run",
+    "--branch=stable",
+    "--arch=x86_64",
+    "--command=ffdec.sh",
+    "com.jpexs.decompiler.flash",
+]
 
-    # wow64 install location
-    if (set_jpexs_if_exists(f"{os.getenv('ProgramFiles(x86)')}\\FFDec\\ffdec.exe")):
-        return True
 
-    return False
+class JPEXSInterface:
+    """An interface to interact with JPEXS via the shell."""
 
-def perror(mesg):
-    print(mesg, file=sys.stderr)
+    path: Path
+    args: list
 
-def read_from_file(file_location, patch_file, current_line_no):
-    try:
-        with open(file_location) as f:
-            current_file = f.readlines()
-            return current_file
-    except (FileNotFoundError, IsADirectoryError) as e:
-        perror("")
-        perror(patch_file + ", line " + str(current_line_no) + ": Invalid injection location")
-        perror("Could not find or load SWF decompiled file at: " + file_location)
-        perror("Aborting...")
-        exit(1)
-
-def write_to_file(path, lines):
-    with open(path, "w") as f:
-        f.writelines(lines)
+    def __init__(self: JPEXSInterface, path: Path, args: list | None = None) -> None:
+        self.path = path
+        if args is None:
+            self.args = []
+        else:
+            self.args = args
+
+    def dump_xml(
+        self: JPEXSInterface,
+        inputfile: Path,
+        output_dir: Path,
+    ) -> subprocess.CompletedProcess:
+        """Dump XML data of the input file into the output location."""
```

Comment:
replace single line comments with `#` where possible

Reply:
This is a docstring, the convention is to use """.